src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,210 @@
+    NEW_FILE,
+    CHUNK_HEAD,
+    CHUNK_BODY,
+    UNKNOWN
+  }
+
+  private static LineType getLineType(String line, boolean isReadingChunk) {
+    if (!isReadingChunk && line.startsWith("---")) {
+      return LineType.OLD_FILE;
+    }
+    if (!isReadingChunk && line.startsWith("+++")) {
+      return LineType.NEW_FILE;
+    }
+    if (line.startsWith("@@") && line.lastIndexOf("@@") != 0) {
+      return LineType.CHUNK_HEAD;
+    }
+    if (isReadingChunk && (line.startsWith("-") || line.startsWith("+") || line.startsWith(" ")
+        || line.isEmpty())) {
+      return LineType.CHUNK_BODY;
+    }
+    return LineType.UNKNOWN;
+  }
+
+  private static List<String> readFile(String file) throws IOException {
+    return Files.readAllLines(new File(file).toPath());
+  }
+
+  private static void writeFile(String file, List<String> content) throws IOException {
+    BufferedWriter writer = new BufferedWriter(new FileWriter(file));
+    for (String line : content) {
+      writer.write(line);
+      writer.write("\n");
+    }
+    writer.close();
+  }
+
+  private static void applyPatchToFile(List<String> patchContent, String oldFile, String newfile)
+      throws IOException, PatchFailedException {
+    List<String> originContent = readFile(oldFile);
+    Patch<String> patch = DiffUtils.parseUnifiedDiff(patchContent);
+    List<String> newContent = DiffUtils.patch(originContent, patch);
+    writeFile(newfile, newContent);
+  }
+
+  private static boolean isGitFormatPatch(List<String> patches) {
+    if (patches.size() > 4) {
+      return patches.get(0).startsWith("From")
+          && patches.get(1).startsWith("From:")
+          && patches.get(2).startsWith("Date:")
+          && patches.get(3).startsWith("Subject:");
+    }
+    return false;
+  }
+
+  /**
+   * Preprocess the original patch file.
+   * 1. Removing leading and trailing empty lines.
+   * 2. Removing ambiguous lines when dealing with git format patch.
+   * 3. Adding an extra line to make sure last chunk also gets applied later.
+   * @param patches The original patch file content
+   * @return The processed patch file content
+   */
+  private static List<String> preprocessPatch(List<String> patches) {
+    // Trimming empty lines from head and tail
+    int left = 0;
+    int right = patches.size() - 1;
+    while (left < patches.size() && patches.get(left).isEmpty()) {
+      left++;
+    }
+    while (right > 0 && patches.get(right).isEmpty()) {
+      right--;
+    }
+    if (left <= right) {
+      patches = patches.subList(left, right + 1);
+    } else {
+      patches = new ArrayList<>();
+    }
+
+    // If it's a git format patch, "--" is ambiguous. We don't know it belongs to a patch chunk or
+    // just the footer of the patch file.
+    // So removing the last 2 lines that looks like:
+    // --
+    // 2.18.0.rc1.244.gcf134e6275-goog
+    if (isGitFormatPatch(patches)) {
+      int size = patches.size();
+      if (patches.get(size - 2).trim().equals("--")) {
```

Comment:
Again,  git patches are just emails; and after the `-- ` (minus-sign, minus-sign, space) on a single line, an arbitrary footer may follow. However, as the footer usually isn't set this might not be so much of a problem.

Reply:
I found this is problematic because `-- ` itself is ambiguous. It could also mean removing the last line which is `- `. Is there any better way to check if there is a footer or not?

Reply:
> I found this is problematic because `-- ` itself is ambiguous. It could also mean removing the last line which is `- `.

Indeed, the footer is not the only place where a line with `-- ` can occur; however, if you parse from the beginning, you always know, how many lines you are expecting in the current chunk.

Reply:
That is smart, let me implement it!